src/server/registry/catalog.go

```diff
@@ -68,16 +74,54 @@ func (r *repositoryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
 		r.sendResponse(w, req, repoNames)
 		return
 	}
-	for _, repo := range repoRecords {
-		valid, err := r.validateRepo(req.Context(), repo.RepositoryID)
-		if err != nil {
-			lib_http.SendError(w, err)
-			return
-		}
-		if valid {
-			repoNames = append(repoNames, repo.Name)
+
+	repoNamesChan := make(chan string)
+	done := make(chan bool, 1)
+	wg.Add(len(repoRecords))
+
+	go func() {
+		defer func() {
+			done <- true
+		}()
+
+		for {
+			select {
+			case str, ok := <-repoNamesChan:
+				if !ok {
+					return
+				}
+				repoNames = append(repoNames, str)
+			}
 		}
+	}()
+
+	for _, repo := range repoRecords {
+		r.wp.GetWorker()
+		go func(repo *models.RepoRecord) {
+			defer func() {
+				r.wp.ReleaseWorker()
+				wg.Done()
+			}()
+
+			tags, err := r.tagCtl.List(orm.Clone(req.Context()), &q.Query{
+				Keywords: map[string]interface{}{
+					"repository_id": repo.RepositoryID,
+				},
+			}, nil)
+			if err != nil {
+				lib_http.SendError(w, err)
```

Comment:
Need to pass the error out of the goroutine and cancel all other goroutines

Reply:
updated